Add S3RemoteStorageManager

ivanyu opened the pull request:
This commit adds the initial version of `S3RemoteStorageManager` code.

ivanyu committed:
Add S3RemoteStorageManager

This commit adds the initial version of `S3RemoteStorageManager` code.

ying-zheng commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/GatheringByteBufferInputStream.java:
```diff
@@ -0,0 +1,74 @@
+
+import org.apache.kafka.common.utils.ByteBufferInputStream;
+
+class GatheringByteBufferInputStream extends InputStream {
```
@ivanyu Can you add a brief introduction (java doc) for each class? That will make the code much more readable

ivanyu replied:
I documented classes and refactored the code to improve readability.

ivanyu committed:
Refactor and document

ivanyu committed:
Fix style and tests

satishd commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java:
```diff
@@ -0,0 +1,659 @@
+ * <p>Sometimes two brokers can think of themselves to be the leader of a partition.
+ * It's also not impossible for them to simultaneously try to upload files from a log segment.
+ * To break this tie, each file uploaded to the remote tier is suffixed with the leader epoch number.
+ * During the read time, the file set with the lower leader epoch number will be used.
```
the file set with the <s>lower</s> **higher** leader epoch number will be used.

ivanyu replied:
@satishd is there any particular reason to pick the highest over the lowest? The lowest feel safer to me since it never changes (the highest can grow).

satishd replied:
@ivanyu A higher leader epoch broker is safer when that common (sub)segment was uploaded with different leader epochs. It is always good to return records with higher leader epochs to avoid any scenarios involving zombie broker(which should have lower leader epoch as some other broker has become leader) thinking itself as the leader.

ivanyu replied:
@satishd got it, changed to use the highest leader epoch.

satishd commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java:
```diff
@@ -0,0 +1,659 @@
+        }
+
+        // No need to explicitly sort the result on our side, we rely on S3 returning in the lexicographical older.
+        return result;
```
No filtering is done here to return only the latest leader epoch segments to the caller. Do you plan to address this later?

ivanyu replied:
@satishd currently, it's filtered by the lowest (implicitly; due to `seenOffsetPairs` set). Let's resolve https://github.com/harshach/kafka/pull/18#discussion_r341111282 and I'll address this accordingly.

satishd commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java:
```diff
@@ -0,0 +1,659 @@
+            throw new KafkaException("Error deleting " + topicPartition, e);
+        }
+
+        return false; // TODO what to return?
```
Right, we need to have a better contract here. We may add IOException to the method signature and avoid any return type. Returning without Exception can be considered as successful. We will address this in `RemoteStorageManager` API.

satishd commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java:
```diff
@@ -0,0 +1,659 @@
+
+            ListObjectsV2Result listObjectsResult;
+            do {
+                // TODO validate this from the read-after-delete point of view
```
Right, this may cause issues if a topic is created after delete. This being addressed in Kafka core by attaching UUID for a topic instead of depending upon topic name. This will allow us to use UUID instead of topic names. We will refactor the APIs once the mentioned enhancement is available.

https://cwiki.apache.org/confluence/display/KAFKA/KIP-516%3A+Topic+Identifiers

ivanyu replied:
Thank you @satishd for pointing out this KIP. I added the link to the comment, but probably nothing to do about it as of now.

satishd commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java:
```diff
@@ -0,0 +1,659 @@
+                        int maxBytes,
+                        long startOffset,
+                        boolean minOneMessage) throws IOException {
+        // TODO what to return in case nothing exists for this offset?
```
Good point, I guess we should throw  `OffsetOutOfRangeException`. WDYT?

ivanyu replied:
@satishd `OffsetOutOfRangeException` seems to be a client exception. We can create a new exception type for this or stick to returning `MemoryRecords.EMPTY`.

satishd reviewed (commented):
Thanks @ivanyu for the PR. Overall LGTM. Left a few minor comments. 
I could not comment on S3 related APIs as I have limited knowledge about them.

ivanyu commented:
Thank you for comments. I'll address them soon!

ivanyu committed:
Merge remote-tracking branch 'harshach/tiered-storage' into tiered-storage-3

# Conflicts:
#	build.gradle

ivanyu committed:
Use minOffset instead of minBaseOffset in listRemoteSegments

This change required switching S3 key names from having the base offset first to having the last offset first.

ivanyu committed:
Use named groups in regexps

ivanyu committed:
Add a comment about KIP-516

ivanyu committed:
Merge remote-tracking branch 'harshach/tiered-storage' into tiered-storage-3

ivanyu committed:
Merge remote-tracking branch 'harshach/tiered-storage' into tiered-storage-3

ivanyu committed:
Use segments with highest leader epoch instead of lowest

ivanyu commented:
Closing in favor of the new implementation based on the new RSM interface.
